### ISSUE TITLE
[Tooling] E2E Automated CLI Testing

### ISSUE THREAD

**Issue by Olshansk** [pokt-network/pocket#466]
## Objective

Automate **CLI** LocalNet E2E and set the foundation for future E2E automation.

## Origin Document

We recently added E2E test automation in [pocket-core/pull/1514/files](https://github.com/pokt-network/pocket-core/pull/1514/files) using [godog](https://github.com/cucumber/godog), which leverages [Cucumber](https://cucumber.io/)'s [Gherkin](https://cucumber.io/docs/gherkin/) syntax.

The framework cannot be copied verbatim over to V1, since a lot of the underlying logic is completely separate, a similar approach can likely be ported over.

We should be able to copy over most of the `README` (with obvious changes) and use the `.go` files as templates, but the problem is relatively open-ended otherwise.

[E2E Testing Framework Design Doc](https://www.notion.so/pocketnetwork/E2E-Testing-Framework-8e8427c8554a49dc9d1bc27bc5772497?pvs=4)

_IMPORTANT NOTE: The `Pacemaker` will need to be in [automatic mode](https://github.com/pokt-network/pocket/blob/main/consensus/debugging.go#L21) to enable this, which can be triggered via the debug command found [here](https://github.com/pokt-network/pocket/blob/main/app/client/cli/debug.go)._

## Goals

- Create the foundation for a framework on which ALL future E2E automation will be built (e.g. including relays, state sync, etc...)
- Enable both application-level automation (e.g. sending a transaction or staking an actor)
- Determine if there is a need/benefit in using https://github.com/testcontainers/testcontainers-go

The following screenshot shows which actions should be automatable by this:

![Screenshot 2023-01-31 at 2 56 34 PM](https://user-images.githubusercontent.com/1892194/215903780-037fcd51-3686-4bd1-bb45-c95db231c780.png)

## Deliverable

- [ ] Migration of the documentation and some code from [pocket-core/e2e](https://github.com/pokt-network/pocket-core/tree/staging/e2e)
- [ ] Automation of the iteration 3 demo found [here](https://github.com/pokt-network/pocket/blob/main/docs/demos/iteration_3_end_to_end_tx.md)

## Non-goals / Non-deliverables

- Adding, removing or changing any core functionality

## General issue deliverables

- [ ] Update the appropriate CHANGELOG(s)
- [ ] Update any relevant local/global README(s)
- [ ] Update relevant source code tree explanations
- [ ] Add or update any relevant or supporting [mermaid](https://mermaid-js.github.io/mermaid/) diagrams

## Testing Methodology

- [ ] **All tests**: `make test_all`
- [ ] **E2E tests**: `make test_e2e` // This is a new addition that should run all e2e_tests at once
- [ ] **E2E stake action**: `make test_e2e_stake_validator` // This is a new addition that should stake a validator
- [ ] **E2E transaction**: `make test_e2e_send_tx` // This is a new addition that should send a transaction
- [ ] A video showing how the automated tests work

---

**Creator**: @Olshansk
**Co-owners**: @jessicadaugherty

**Comment by Olshansk** [comment]
@h5law I know you already have a lot on your plate, but just wanted to share this as an FYI to insert a mental tab (to think about for the future) as you'll be working on integrating the keybase into our CLI. 💭 

The pocket-core PR can act as a starting reference, but there's lots of new tooling and automation here in case you're up for the challenge? 🤔 

cc @jessicadaugherty

**Comment by jessicadaugherty** [comment]
Thanks @Olshansk! Do you think there's an opportunity to split this ticket by tx and localnet debug? If not, would you size this as an XL?

**Comment by Olshansk** [comment]
@jessicadaugherty Great question!

I ended up building #481 while trying to get an answer to your question. 

I have updated the requirements (with extra details) here, and created #482 as a followup to this focused primarily on debug. My aim is that the code & documentation in v0 can help guide the development of this ticket and #481 can help guide the development of #482, though I still believe some questions during the dev cycle make come up.

I've marked the effort here as XXL and the followup ticket as XL.

**Comment by jessicadaugherty** [comment]
Wow, very generous and much appreciated! Thank you, I will get this sorted out on the bounty board. Tysm :)

**Comment by h5law** [comment]
@Olshansk Yeah sure! I will first work on the keybase integration with the CLI - (not the endpoints but in place of the current key usages)

Once this is done I will begin to look into the E2E testing stuff more. I checkout out the `pocket-core` e2e tests with godog and will try to learn it all as best I can 🚀

**Comment by jessicadaugherty** [comment]
To be broken down into smaller issues

**Comment by jessicadaugherty** [comment]
Add design document here @dylanlott

**Comment by dylanlott** [comment]
> Add design document here @dylanlott 
[E2E Testing Framework](https://www.notion.so/pocketnetwork/E2E-Testing-Framework-8e8427c8554a49dc9d1bc27bc5772497)
